In [50]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

import os

print('Libraries imported.')

Libraries imported.


# 1. Download and Explore the dataset

In [51]:
os.chdir('D:\Data Science\IBM Capstone Project')
pd.read_json('nyu-geojson.json', lines = True)

bbox  \
0  [-74.2492599487305, 40.5033187866211, -73.7061...   

                                                 crs  \
0  {'type': 'name', 'properties': {'name': 'urn:o...   

                                            features  totalFeatures  \
0  [{'type': 'Feature', 'id': 'nyu_2451_34572.1',...            306   

                type  
0  FeatureCollection

In [52]:
#Reading the json file as a dict
with open("nyu-geojson.json") as datafile:
    data1 = json.load(datafile)

In [53]:
data1

{'bbox': [-74.2492599487305,
  40.5033187866211,
  -73.7061614990234,
  40.9105606079102],
 'crs': {'properties': {'name': 'urn:ogc:def:crs:EPSG::4326'}, 'type': 'name'},
 'features': [{'geometry': {'coordinates': [-73.84720052054902,
     40.89470517661],
    'type': 'Point'},
   'geometry_name': 'geom',
   'id': 'nyu_2451_34572.1',
   'properties': {'annoangle': 0.0,
    'annoline1': 'Wakefield',
    'annoline2': None,
    'annoline3': None,
    'bbox': [-73.84720052054902,
     40.89470517661,
     -73.84720052054902,
     40.89470517661],
    'borough': 'Bronx',
    'name': 'Wakefield',
    'stacked': 1},
   'type': 'Feature'},
  {'geometry': {'coordinates': [-73.82993910812398, 40.87429419303012],
    'type': 'Point'},
   'geometry_name': 'geom',
   'id': 'nyu_2451_34572.2',
   'properties': {'annoangle': 0.0,
    'annoline1': 'Co-op',
    'annoline2': 'City',
    'annoline3': None,
    'bbox': [-73.82993910812398,
     40.87429419303012,
     -73.82993910812398,
     40.874294193

In [54]:
neighborhoods_data = data1['features']

Let's take a look at the first item in this list.

In [55]:
neighborhoods_data[0]

{'geometry': {'coordinates': [-73.84720052054902, 40.89470517661],
  'type': 'Point'},
 'geometry_name': 'geom',
 'id': 'nyu_2451_34572.1',
 'properties': {'annoangle': 0.0,
  'annoline1': 'Wakefield',
  'annoline2': None,
  'annoline3': None,
  'bbox': [-73.84720052054902,
   40.89470517661,
   -73.84720052054902,
   40.89470517661],
  'borough': 'Bronx',
  'name': 'Wakefield',
  'stacked': 1},
 'type': 'Feature'}

#### Tranform the data into a *pandas* dataframe

The next task is essentially transforming this data of nested Python dictionaries into a pandas dataframe. So let's start by creating an empty dataframe.

In [56]:
# define the dataframe columns
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 

In [57]:
# instantiate the dataframe
neighborhoods = pd.DataFrame(columns=column_names)

Take a look at the empty dataframe to confirm that the columns are as intended.

In [58]:
neighborhoods

Empty DataFrame
Columns: [Borough, Neighborhood, Latitude, Longitude]
Index: []

Then let's loop through the data and fill the dataframe one row at a time.

In [59]:
for data in neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough']
    neighborhood_name = data['properties']['name']
    
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    neighborhoods = neighborhoods.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)

In [60]:
neighborhoods.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

And make sure that the dataset has all 5 boroughs and 306 neighborhoods.

In [61]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]
    )
)

The dataframe has 5 boroughs and 306 neighborhoods.


### Use geopy library to get the latitude and longitude values of New York City.
In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent ny_explorer, as shown below.

In [62]:
address = 'New York City, NY'

geolocator = Nominatim(user_agent='ny_explorer')
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of New York City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of New York City are 40.7127281, -74.0060152.


### Create a map of New York with neighborhoods superimposed on top.

In [63]:
# create map of New York using latitude and longitude values
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

In [64]:
# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)
    
map_newyork

However, for illustration purposes, let's simplify the above map and segment and cluster only the neighborhoods in Manhattan. So let's slice the original dataframe and create a new dataframe of the Manhattan data.

In [69]:
manhattan_data = neighborhoods[neighborhoods['Borough'] == 'Manhattan'].reset_index(drop=True)

In [70]:
manhattan_data.head()

Borough        Neighborhood   Latitude  Longitude
0  Manhattan         Marble Hill  40.876551 -73.910660
1  Manhattan           Chinatown  40.715618 -73.994279
2  Manhattan  Washington Heights  40.851903 -73.936900
3  Manhattan              Inwood  40.867684 -73.921210
4  Manhattan    Hamilton Heights  40.823604 -73.949688

Let's get the geographical coordinates of Manhattan.

In [71]:
address = 'Manhattan, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Manhattan are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Manhattan are 40.7900869, -73.9598295.


As we did with all of New York City, let's visualize Manhattan the neighborhoods in it.

In [72]:
# create map of Manhattan using latitude and longitude values
map_manhattan = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(manhattan_data['Latitude'], manhattan_data['Longitude'], manhattan_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_manhattan)  
    
map_manhattan

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them

In [73]:
CLIENT_ID = 'CVNOQOLPHNEJIEJIVI1YIULE2UYGLBFV0GE01URDRJRRI430' # your Foursquare ID
CLIENT_SECRET = 'IGM3THDU2LIGW105V2SYPYQ1ZSYG3ZAJFTPLPS2PTEZLSPYB' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: CVNOQOLPHNEJIEJIVI1YIULE2UYGLBFV0GE01URDRJRRI430
CLIENT_SECRET:IGM3THDU2LIGW105V2SYPYQ1ZSYG3ZAJFTPLPS2PTEZLSPYB


<b>Let's explore the first neighborhood in our dataframe.

Get the neighborhood's name.

In [74]:
manhattan_data.Neighborhood[0]

'Marble Hill'

Get the neighborhood's latitude and longitude values.

In [75]:
neighborhood_latitude = manhattan_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = manhattan_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = manhattan_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Marble Hill are 40.87655077879964, -73.91065965862981.


<b>Now, let's get the top 100 venues that are in Marble Hill within a radius of 500 meters

First, let's create the GET request URL. Name your URL url

In [76]:
LIMIT = 100

radius = 500

#create url
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=CVNOQOLPHNEJIEJIVI1YIULE2UYGLBFV0GE01URDRJRRI430&client_secret=IGM3THDU2LIGW105V2SYPYQ1ZSYG3ZAJFTPLPS2PTEZLSPYB&v=20180605&ll=40.87655077879964,-73.91065965862981&radius=500&limit=100'

Send the <b>GET request</b> and examine the results

In [77]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5cf956db6a607149399b56fb'},
 'response': {'groups': [{'items': [{'reasons': {'count': 0,
       'items': [{'reasonName': 'globalInteractionReason',
         'summary': 'This spot is popular',
         'type': 'general'}]},
      'referralId': 'e-0-4b4429abf964a52037f225e3-0',
      'venue': {'categories': [{'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/pizza_',
          'suffix': '.png'},
         'id': '4bf58dd8d48988d1ca941735',
         'name': 'Pizza Place',
         'pluralName': 'Pizza Places',
         'primary': True,
         'shortName': 'Pizza'}],
       'delivery': {'id': '72548',
        'provider': {'icon': {'name': '/delivery_provider_seamless_20180129.png',
          'prefix': 'https://fastly.4sqi.net/img/general/cap/',
          'sizes': [40, 50]},
         'name': 'seamless'},
        'url': 'https://www.seamless.com/menu/arturos-pizza-5189-broadway-ave-new-york/72548?affiliate=1131&utm_source=foursquare-affiliat

From the <b>Foursquare lab</b> in the previous module, we know that all the information is in the items key. Before we proceed, let's borrow the <b>get_category_type</b> function from the Foursquare lab.

In [78]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the <b>json</b> and structure it into a pandas dataframe.

## explain

In [80]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name   categories        lat        lng
0       Arturo's  Pizza Place  40.874412 -73.910271
1    Bikram Yoga  Yoga Studio  40.876844 -73.906204
2  Tibbett Diner        Diner  40.880404 -73.908937
3        Dunkin'   Donut Shop  40.877136 -73.906666
4      Starbucks  Coffee Shop  40.877531 -73.905582

And how many venues were returned by Foursquare?

In [81]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

25 venues were returned by Foursquare.


# 2. Explore Neighborhoods in Manhattan

<b>Let's create a function to repeat the same process to all the neighborhoods in Manhattan

In [82]:
def getNearbyVenues(names, latitudes, longitudes, radius = 500):
    venues_list = []
    
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
        
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
        
        # make the GET request
        
        results1 = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results1])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

### Now write the code to run the above function on each neighborhood and create a new dataframe called manhattan_venues.

In [83]:
manhattan_venues = getNearbyVenues(names = manhattan_data['Neighborhood'], latitudes = manhattan_data['Latitude'], longitudes = manhattan_data['Longitude'])

Marble Hill
Chinatown
Washington Heights
Inwood
Hamilton Heights
Manhattanville
Central Harlem
East Harlem
Upper East Side
Yorkville
Lenox Hill
Roosevelt Island
Upper West Side
Lincoln Square
Clinton
Midtown
Murray Hill
Chelsea
Greenwich Village
East Village
Lower East Side
Tribeca
Little Italy
Soho
West Village
Manhattan Valley
Morningside Heights
Gramercy
Battery Park City
Financial District
Carnegie Hill
Noho
Civic Center
Midtown South
Sutton Place
Turtle Bay
Tudor City
Stuyvesant Town
Flatiron
Hudson Yards


In [84]:
print(manhattan_venues.shape)
manhattan_venues.head()

(3330, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude          Venue  \
0  Marble Hill              40.876551               -73.91066       Arturo's   
1  Marble Hill              40.876551               -73.91066    Bikram Yoga   
2  Marble Hill              40.876551               -73.91066  Tibbett Diner   
3  Marble Hill              40.876551               -73.91066        Dunkin'   
4  Marble Hill              40.876551               -73.91066      Starbucks   

   Venue Latitude  Venue Longitude Venue Category  
0       40.874412       -73.910271    Pizza Place  
1       40.876844       -73.906204    Yoga Studio  
2       40.880404       -73.908937          Diner  
3       40.877136       -73.906666     Donut Shop  
4       40.877531       -73.905582    Coffee Shop

Let's check how many venues were returned for each neighborhood

In [85]:
manhattan_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  \
Neighborhood                                                                
Battery Park City                      100                     100    100   
Carnegie Hill                          100                     100    100   
Central Harlem                          47                      47     47   
Chelsea                                100                     100    100   
Chinatown                              100                     100    100   
Civic Center                           100                     100    100   
Clinton                                100                     100    100   
East Harlem                             44                      44     44   
East Village                           100                     100    100   
Financial District                     100                     100    100   
Flatiron                               100                     100    100   
Gramercy                               100                     100    100   
Greenwich Village                      100                     100    100   
Hamilton Heights                        62                      62     62   
Hudson Yards                            75                      75     75   
Inwood                                  58                      58     58   
Lenox Hill                             100                     100    100   
Lincoln Square                         100                     100    100   
Little Italy                           100                     100    100   
Lower East Side                         60                      60     60   
Manhattan Valley                        59                      59     59   
Manhattanville                          44                      44     44   
Marble Hill                             25                      25     25   
Midtown                                100                     100    100   
Midtown South                          100                     100    100   
Morningside Heights                     40                      40     40   
Murray Hill                            100                     100    100   
Noho                                   100                     100    100   
Roosevelt Island                        33                      33     33   
Soho                                   100                     100    100   
Stuyvesant Town                         16                      16     16   
Sutton Place                           100                     100    100   
Tribeca                                100                     100    100   
Tudor City                              82                      82     82   
Turtle Bay                             100                     100    100   
Upper East Side                        100                     100    100   
Upper West Side                        100                     100    100   
Washington Heights                      85                      85     85   
West Village                           100                     100    100   
Yorkville                              100                     100    100   

                     Venue Latitude  Venue Longitude  Venue Category  
Neighborhood                                                          
Battery Park City               100              100             100  
Carnegie Hill                   100              100             100  
Central Harlem                   47               47              47  
Chelsea                         100              100             100  
Chinatown                       100              100             100  
Civic Center                    100              100             100  
Clinton                         100              100             100  
East Harlem                      44               44              44  
East Village                    100              100             100  
Finan

<b>Let's find out how many unique categories can be curated from all the returned venues

In [86]:
print('There are {} uniques categories.'.format(len(manhattan_venues['Venue Category'].unique())))

There are 333 uniques categories.


# 3. Analyze Each Neighborhood

In [87]:
# one hot encoding
manhattan_onehot = pd.get_dummies(manhattan_venues[['Venue Category']], prefix = "", prefix_sep = "")

# add neighborhood column back to dataframe
manhattan_onehot['Neighborhood'] = manhattan_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [manhattan_onehot.columns[-1]] + list(manhattan_onehot.columns[:-1])
manhattan_onehot = manhattan_onehot[fixed_columns]

manhattan_onehot.head()

Neighborhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0  Marble Hill                  0               0                  0   
1  Marble Hill                  0               0                  0   
2  Marble Hill                  0               0                  0   
3  Marble Hill                  0               0                  0   
4  Marble Hill                  0               0                  0   

   African Restaurant  American Restaurant  Animal Shelter  Antique Shop  \
0                   0                    0               0             0   
1                   0                    0               0             0   
2                   0                    0               0             0   
3                   0                    0               0             0   
4                   0                    0               0             0   

   Arcade  Arepa Restaurant  Argentinian Restaurant  Art Gallery  Art Museum  \
0       0                 0                       0            0           0   
1       0                 0                       0            0           0   
2       0                 0                       0            0           0   
3       0                 0                       0            0           0   
4       0                 0                       0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auditorium  \
0                    0                 0                   0           0   
1                    0                 0                   0           0   
2                    0                 0                   0           0   
3                    0                 0                   0           0   
4                    0                 0                   0           0   

   Australian Restaurant  Austrian Restaurant  Auto Workshop  BBQ Joint  \
0                      0                    0              0          0   
1                      0                    0              0          0   
2                      0                    0              0          0   
3                      0                    0              0          0   
4                      0                    0              0          0   

   Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0           0           0       0     0    0               0   
1           0           0       0     0    0               0   
2           0           0       0     0    0               0   
3           0           0       0     0    0               0   
4           0           0       0     0    0               0   

   Basketball Court  Beer Bar  Beer Garden  Beer Store  Big Box Store  \
0                 0         0            0           0              0   
1                 0         0            0           0              0   
2                 0         0            0           0              0   
3                 0         0            0           0              0   
4                 0         0            0           0              0   

   Bike Rental / Bike Share  Bike Shop  Bike Trail  Bistro  Board Shop  \
0                         0          0           0       0           0   
1                         0          0           0       0           0   
2                         0          0           0       0           0   
3                         0          0           0       0           0   
4                         0          0           0       0           0   

   Boat or Ferry  Bookstore  Boutique  Boxing Gym  Brazilian Restaurant  \
0              0          0         0           0                     0   
1              0          0         0           0                     0   
2              0          0         0           0                     0   
3              0          0         0           0                     0   
4              0          0         0           0                     0   

   

And let's examine the new dataframe size.

In [88]:
manhattan_onehot.shape

(3330, 334)

<b>Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [89]:
manhattan_grouped = manhattan_onehot.groupby('Neighborhood').mean()

In [90]:
manhattan_grouped.head()

Accessories Store  Adult Boutique  Afghan Restaurant  \
Neighborhood                                                              
Battery Park City                0.0             0.0                0.0   
Carnegie Hill                    0.0             0.0                0.0   
Central Harlem                   0.0             0.0                0.0   
Chelsea                          0.0             0.0                0.0   
Chinatown                        0.0             0.0                0.0   

                   African Restaurant  American Restaurant  Animal Shelter  \
Neighborhood                                                                 
Battery Park City             0.00000             0.010000             0.0   
Carnegie Hill                 0.00000             0.010000             0.0   
Central Harlem                0.06383             0.042553             0.0   
Chelsea                       0.00000             0.040000             0.0   
Chinatown                     0.00000             0.040000             0.0   

                   Antique Shop  Arcade  Arepa Restaurant  \
Neighborhood                                                
Battery Park City          0.00     0.0               0.0   
Carnegie Hill              0.00     0.0               0.0   
Central Harlem             0.00     0.0               0.0   
Chelsea                    0.01     0.0               0.0   
Chinatown                  0.00     0.0               0.0   

                   Argentinian Restaurant  Art Gallery  Art Museum  \
Neighborhood                                                         
Battery Park City                     0.0     0.000000        0.00   
Carnegie Hill                         0.0     0.000000        0.01   
Central Harlem                        0.0     0.042553        0.00   
Chelsea                               0.0     0.020000        0.00   
Chinatown                             0.0     0.000000        0.00   

                   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
Neighborhood                                                                   
Battery Park City                  0.0              0.00                0.01   
Carnegie Hill                      0.0              0.00                0.00   
Central Harlem                     0.0              0.00                0.00   
Chelsea                            0.0              0.01                0.00   
Chinatown                          0.0              0.02                0.00   

                   Auditorium  Australian Restaurant  Austrian Restaurant  \
Neighborhood                                                                
Battery Park City        0.01                    0.0                 0.00   
Carnegie Hill            0.00                    0.0                 0.00   
Central Harlem           0.00                    0.0                 0.00   
Chelsea                  0.00                    0.0                 0.00   
Chinatown                0.00                    0.0                 0.01   

                   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop    Bakery  \
Neighborhood                                                                    
Battery Park City            0.0   0.020000         0.0    0.000000  0.010000   
Carnegie Hill                0.0   0.000000         0.0    0.010000  0.020000   
Central Harlem               0.0   0.021277         0.0    0.021277  0.021277   
Chelsea                      0.0   0.000000         0.0    0.010000  0.040000   
Chinatown                    0.0   0.000000         0.0    0.000000  0.020000   

                   Bank       Bar  Baseball Field  Basketball Court  Beer Bar  \
Neighborhood                                                                    
Battery Park City   0.0  0.000000             0.0               0.0  0.000000   
Carnegie Hill       0.0  0.030000             0.0               0.0  0.000000   
Central Harlem      0.0  0.021277 

In [91]:
manhattan_grouped.reset_index(inplace = True)

In [93]:
manhattan_grouped.head()

Neighborhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0  Battery Park City                0.0             0.0                0.0   
1      Carnegie Hill                0.0             0.0                0.0   
2     Central Harlem                0.0             0.0                0.0   
3            Chelsea                0.0             0.0                0.0   
4          Chinatown                0.0             0.0                0.0   

   African Restaurant  American Restaurant  Animal Shelter  Antique Shop  \
0             0.00000             0.010000             0.0          0.00   
1             0.00000             0.010000             0.0          0.00   
2             0.06383             0.042553             0.0          0.00   
3             0.00000             0.040000             0.0          0.01   
4             0.00000             0.040000             0.0          0.00   

   Arcade  Arepa Restaurant  Argentinian Restaurant  Art Gallery  Art Museum  \
0     0.0               0.0                     0.0     0.000000        0.00   
1     0.0               0.0                     0.0     0.000000        0.01   
2     0.0               0.0                     0.0     0.042553        0.00   
3     0.0               0.0                     0.0     0.020000        0.00   
4     0.0               0.0                     0.0     0.000000        0.00   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auditorium  \
0                  0.0              0.00                0.01        0.01   
1                  0.0              0.00                0.00        0.00   
2                  0.0              0.00                0.00        0.00   
3                  0.0              0.01                0.00        0.00   
4                  0.0              0.02                0.00        0.00   

   Australian Restaurant  Austrian Restaurant  Auto Workshop  BBQ Joint  \
0                    0.0                 0.00            0.0   0.020000   
1                    0.0                 0.00            0.0   0.000000   
2                    0.0                 0.00            0.0   0.021277   
3                    0.0                 0.00            0.0   0.000000   
4                    0.0                 0.01            0.0   0.000000   

   Baby Store  Bagel Shop    Bakery  Bank       Bar  Baseball Field  \
0         0.0    0.000000  0.010000   0.0  0.000000             0.0   
1         0.0    0.010000  0.020000   0.0  0.030000             0.0   
2         0.0    0.021277  0.021277   0.0  0.021277             0.0   
3         0.0    0.010000  0.040000   0.0  0.010000             0.0   
4         0.0    0.000000  0.020000   0.0  0.010000             0.0   

   Basketball Court  Beer Bar  Beer Garden  Beer Store  Big Box Store  \
0               0.0  0.000000         0.01         0.0            0.0   
1               0.0  0.000000         0.00         0.0            0.0   
2               0.0  0.021277         0.00         0.0            0.0   
3               0.0  0.010000         0.00         0.0            0.0   
4               0.0  0.000000         0.00         0.0            0.0   

   Bike Rental / Bike Share  Bike Shop  Bike Trail  Bistro  Board Shop  \
0                       0.0       0.00         0.0    0.01         0.0   
1                       0.0       0.00         0.0    0.00         0.0   
2                       0.0       0.00         0.0    0.00         0.0   
3                       0.0       0.00         0.0    0.00         0.0   
4                       0.0       0.01         0.0    0.00         0.0   

   Boat or Ferry  Bookstore  Boutique  Boxing Gym  Brazilian Restaurant  \
0           0.01   0.000000  0.000000         0.0                   0.0   
1           0.00   0.030000  0.000000         0.0                   0.0   
2           0.00   0.021277  0.021277         0.0                   0.0   
3           0.00   0.020000  0.000000         0.0                   0.0   
4      

Let's confirm the new size

In [94]:
manhattan_grouped.shape

(40, 334)

<b>Let's print each neighborhood along with the top 5 most common venues

In [97]:
num_top_venues = 5

for hood in manhattan_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = manhattan_grouped[manhattan_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Battery Park City----
           venue  freq
0           Park  0.08
1    Coffee Shop  0.07
2          Hotel  0.05
3  Memorial Site  0.04
4            Gym  0.03


----Carnegie Hill----
                 venue  freq
0          Pizza Place  0.06
1          Coffee Shop  0.06
2                 Café  0.04
3  Japanese Restaurant  0.03
4    French Restaurant  0.03


----Central Harlem----
                venue  freq
0  African Restaurant  0.06
1  Seafood Restaurant  0.04
2         Art Gallery  0.04
3      Cosmetics Shop  0.04
4   French Restaurant  0.04


----Chelsea----
                 venue  freq
0          Coffee Shop  0.06
1   Italian Restaurant  0.05
2       Ice Cream Shop  0.05
3               Bakery  0.04
4  American Restaurant  0.04


----Chinatown----
                 venue  freq
0   Chinese Restaurant  0.10
1  American Restaurant  0.04
2         Cocktail Bar  0.04
3  Dumpling Restaurant  0.03
4      Bubble Tea Shop  0.03


----Civic Center----
                  venue  freq
0    I

Let's put that into a pandas dataframe

First, let's write a function to sort the venues in <b>descending order.

In [98]:
def return_most_common_venues(row, num_top_venue):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending = True)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the <b>top 10 venues</b> for each neighborhood.

In [99]:
num_top_ven = 10

indicators = ['st','nd','rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))        ####EXPLAIN


# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = manhattan_grouped['Neighborhood']

for ind in np.arange(manhattan_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(manhattan_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue    2nd Most Common Venue  \
0  Battery Park City     Accessories Store              Opera House   
1      Carnegie Hill     Accessories Store              Opera House   
2     Central Harlem     Accessories Store        Paella Restaurant   
3            Chelsea     Accessories Store  North Indian Restaurant   
4          Chinatown     Accessories Store                   Office   

     3rd Most Common Venue    4th Most Common Venue 5th Most Common Venue  
0                   Office  North Indian Restaurant          Noodle House  
1                   Office  North Indian Restaurant          Noodle House  
2    Outdoors & Recreation        Outdoor Sculpture       Other Nightlife  
3             Noodle House               Non-Profit             Newsstand  
4  North Indian Restaurant               Non-Profit             Nightclub

# 4. Cluster Neighborhoods

Run <b>k-means</b> to cluster the neighborhood into 5 clusters.

In [100]:
# set number of clusters

kclusters = 5

#Preparing data

manhattan_grouped_clustering = manhattan_grouped.drop('Neighborhood', 1)

#Fitting the data

kmeans = KMeans(n_clusters = kclusters, random_state = 0)
kmeans.fit(manhattan_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([3, 1, 1, 0, 0, 0, 0, 4, 0, 0])

Let's create a <b>new dataframe</b> that includes the cluster as well as the <b>top 10 venues</b> for each neighborhood.

In [101]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

manhattan_merged = manhattan_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
manhattan_merged = manhattan_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

manhattan_merged.head() # check the last columns!

Borough        Neighborhood   Latitude  Longitude  Cluster Labels  \
0  Manhattan         Marble Hill  40.876551 -73.910660               4   
1  Manhattan           Chinatown  40.715618 -73.994279               0   
2  Manhattan  Washington Heights  40.851903 -73.936900               4   
3  Manhattan              Inwood  40.867684 -73.921210               4   
4  Manhattan    Hamilton Heights  40.823604 -73.949688               4   

  1st Most Common Venue 2nd Most Common Venue    3rd Most Common Venue  \
0     Accessories Store     Outdoor Sculpture          Other Nightlife   
1     Accessories Store                Office  North Indian Restaurant   
2             Irish Pub             Newsstand               Nail Salon   
3     Accessories Store           Opera House                   Office   
4     Accessories Store       Other Nightlife          Organic Grocery   

     4th Most Common Venue 5th Most Common Venue  
0          Organic Grocery          Optical Shop  
1               Non-Profit             Nightclub  
2              Music Venue          Music School  
3  North Indian Restaurant          Noodle House  
4             Optical Shop           Opera House

Finally, let's visualize the <b>resulting clusters

In [102]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(manhattan_merged['Latitude'], manhattan_merged['Longitude'], manhattan_merged['Neighborhood'], manhattan_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

# 5. Examine Clusters

Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster.

## Cluster 1

In [103]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 0, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue    2nd Most Common Venue  \
1            Chinatown     Accessories Store                   Office   
8      Upper East Side     Accessories Store              Opera House   
13      Lincoln Square     Accessories Store          Organic Grocery   
14             Clinton     Accessories Store                Nightclub   
17             Chelsea     Accessories Store  North Indian Restaurant   
18   Greenwich Village     Accessories Store  North Indian Restaurant   
19        East Village     Accessories Store             Noodle House   
21             Tribeca     Accessories Store  North Indian Restaurant   
24        West Village             Irish Pub             Optical Shop   
27            Gramercy     Accessories Store        Jewish Restaurant   
29  Financial District             Irish Pub          Other Nightlife   
31                Noho     Accessories Store  North Indian Restaurant   
32        Civic Center     Accessories Store        Outdoor Sculpture   
35          Turtle Bay     Accessories Store               Non-Profit   
39        Hudson Yards     Accessories Store              Opera House   

      3rd Most Common Venue    4th Most Common Venue    5th Most Common Venue  
1   North Indian Restaurant               Non-Profit                Nightclub  
8                    Office  North Indian Restaurant             Noodle House  
13                   Office  North Indian Restaurant             Noodle House  
14                Newsstand               Nail Salon              Music Venue  
17             Noodle House               Non-Profit                Newsstand  
18             Noodle House               Non-Profit                Nightclub  
19               Non-Profit                Nightclub                Newsstand  
21               Non-Profit                Nightclub                Newsstand  
24              Opera House                   Office  North Indian Restaurant  
27     Taiwanese Restaurant              Karaoke Bar         Kebab Restaurant  
29          Organic Grocery             Optical Shop              Opera House  
31             Noodle House               Non-Profit                Nightclub  
32          Other Nightlife          Organic Grocery             Optical Shop  
35                Nightclub                Newsstand  New American Restaurant  
39                   Office  North Indian Restaurant             Noodle House

## Cluster 2

In [104]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 1, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue    2nd Most Common Venue  \
6    Central Harlem     Accessories Store        Paella Restaurant   
9         Yorkville     Accessories Store          Organic Grocery   
10       Lenox Hill     Accessories Store             Noodle House   
12  Upper West Side             Irish Pub            Jewelry Store   
15          Midtown     Accessories Store          Organic Grocery   
16      Murray Hill     Accessories Store                Nightclub   
22     Little Italy     Accessories Store  North Indian Restaurant   
23             Soho             Irish Pub         Kebab Restaurant   
30    Carnegie Hill     Accessories Store              Opera House   
33    Midtown South     Accessories Store          Organic Grocery   
34     Sutton Place     Accessories Store               Taco Place   
38         Flatiron     Accessories Store             Optical Shop   

    3rd Most Common Venue    4th Most Common Venue    5th Most Common Venue  
6   Outdoors & Recreation        Outdoor Sculpture          Other Nightlife  
9            Optical Shop              Opera House                   Office  
10              Nightclub                Newsstand  New American Restaurant  
12      Jewish Restaurant         Swiss Restaurant              Karaoke Bar  
15            Opera House                   Office  North Indian Restaurant  
16              Newsstand               Nail Salon              Music Venue  
22             Non-Profit                Nightclub  New American Restaurant  
23             Kids Store        Korean Restaurant        Kosher Restaurant  
30                 Office  North Indian Restaurant             Noodle House  
33           Optical Shop              Opera House                   Office  
34            Karaoke Bar         Kebab Restaurant               Kids Store  
38            Opera House  North Indian Restaurant             Noodle House

## Cluster 3

In [105]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 2, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
37  Stuyvesant Town     Accessories Store     Paella Restaurant   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  
37  Outdoors & Recreation     Outdoor Sculpture       Other Nightlife

## Cluster 4

In [106]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 3, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
11     Roosevelt Island     Accessories Store       Other Nightlife   
26  Morningside Heights     Accessories Store       Other Nightlife   
28    Battery Park City     Accessories Store           Opera House   

   3rd Most Common Venue    4th Most Common Venue 5th Most Common Venue  
11       Organic Grocery             Optical Shop           Opera House  
26       Organic Grocery             Optical Shop           Opera House  
28                Office  North Indian Restaurant          Noodle House

## Cluster 5

In [107]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 4, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  2nd Most Common Venue  \
0          Marble Hill     Accessories Store      Outdoor Sculpture   
2   Washington Heights             Irish Pub              Newsstand   
3               Inwood     Accessories Store            Opera House   
4     Hamilton Heights     Accessories Store        Other Nightlife   
5       Manhattanville     Accessories Store  Outdoors & Recreation   
7          East Harlem     Accessories Store           Optical Shop   
20     Lower East Side     Accessories Store           Optical Shop   
25    Manhattan Valley     Accessories Store      Outdoor Sculpture   
36          Tudor City     Accessories Store                 Office   

      3rd Most Common Venue    4th Most Common Venue    5th Most Common Venue  
0           Other Nightlife          Organic Grocery             Optical Shop  
2                Nail Salon              Music Venue             Music School  
3                    Office  North Indian Restaurant             Noodle House  
4           Organic Grocery             Optical Shop              Opera House  
5         Outdoor Sculpture          Organic Grocery             Optical Shop  
7               Opera House                   Office  North Indian Restaurant  
20              Opera House                   Office  North Indian Restaurant  
25          Other Nightlife          Organic Grocery             Optical Shop  
36  North Indian Restaurant             Noodle House               Non-Profit